### PR TITLE
Ledgerdata

### PR DESCRIPTION
**Pull request by cindyyan317**
"ledger_data" in new RPC framework.
Worth to mention: "out_of_order" field is only in Clio, I just copy the logic from old handler.

### CODE REVIEW

File: src/rpc/ngHandlers/LedgerData.cpp [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+//==============================================================================
+
+#include <rpc/ngHandlers/LedgerData.h>
+
+#include <ripple/app/ledger/LedgerToJson.h>
+
+#include <algorithm>
+// local to compilation unit loggers
+namespace {
+clio::Logger gLog{"RPC"};
```
Comment (godexsoft) [inline src/rpc/ngHandlers/LedgerData.cpp]: this can be a private member now :)

File: src/rpc/ngHandlers/LedgerData.cpp [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+}  // namespace
+
+namespace RPCng {
+
+LedgerDataHandler::Result
+LedgerDataHandler::process(Input input, Context const& ctx) const
+{
+    // marker must be int if outOfOrder is true
+    if (input.outOfOrder && input.marker)
+    {
```
Comment (godexsoft) [inline src/rpc/ngHandlers/LedgerData.cpp]: can remove extra braces here and below

File: src/rpc/ngHandlers/LedgerData.h [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+             validation::IfType<std::string>{validation::Uint256HexStringValidator}},
+        };
+        return rpcSpec;
+    }
+
+    Result
+    process(Input input, Context const& ctx) const;
+
+private:
+    static uint32_t constexpr limitIfBinary = 2048;
```
Comment (godexsoft) [inline src/rpc/ngHandlers/LedgerData.h]: I think we use all caps for constants in other places? let's keep it uniform :)

**Review by godexsoft — CHANGES_REQUESTED** [review]
I know this is still a draft but i thought i'd leave a few comments while i was already here 🫡 Just small stuff, nothing major.

File: src/rpc/ngHandlers/LedgerData.cpp [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    output.ledgerIndex = lgrInfo.seq;
+
+    auto const start = std::chrono::system_clock::now();
+    std::vector<Backend::LedgerObject> results;
+    if (input.diffMarker)
+    {
+        // keep the same logic as previous implementation
+        auto const diff = sharedPtrBackend_->fetchLedgerDiff(*(input.diffMarker), ctx.yield);
+        std::vector<ripple::uint256> keys;
+        for (auto& [key, object] : diff)
```
Comment (cindyyan317) [inline src/rpc/ngHandlers/LedgerData.cpp:92]: Old handler uses " for (auto&& [key, object] : diff)" . change to auto&, auto && is confusing here

File: src/rpc/ngHandlers/LedgerData.cpp [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+        {
+            if (!object.size())
+                keys.push_back(std::move(key));
+        }
+        auto objs = sharedPtrBackend_->fetchLedgerObjects(keys, lgrInfo.seq, ctx.yield);
+        for (size_t i = 0; i < objs.size(); ++i)
+        {
+            auto& obj = objs[i];
```
Comment (cindyyan317) [inline src/rpc/ngHandlers/LedgerData.cpp:100]: change auto&& to auto&

File: src/rpc/ngHandlers/LedgerData.cpp (hunk 4 above)
Comment (godexsoft) [reply src/rpc/ngHandlers/LedgerData.cpp:92]: can you check if 'auto const diff' above breaks move below?

File: src/rpc/ngHandlers/LedgerData.cpp [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+        auto page = sharedPtrBackend_->fetchLedgerPage(input.marker, lgrInfo.seq, limit, input.outOfOrder, ctx.yield);
+        results = std::move(page.objects);
+        if (page.cursor)
+            output.marker = ripple::strHex(*(page.cursor));
+        else if (input.outOfOrder)
+            output.diffMarker = sharedPtrBackend_->fetchLedgerRange()->maxSequence;
+    }
+    auto const end = std::chrono::system_clock::now();
+
+    auto time = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();
```
Comment (godexsoft) [inline src/rpc/ngHandlers/LedgerData.cpp]: can be const?

File: src/rpc/ngHandlers/LedgerData.cpp [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+        else
+        {
+            output.states.push_back(RPC::toJson(sle));
+        }
+    }
+    if (input.outOfOrder)
+        output.cacheFull = sharedPtrBackend_->cache().isFull();
+    auto const end2 = std::chrono::system_clock::now();
+
+    time = std::chrono::duration_cast<std::chrono::microseconds>(end2 - end).count();
```
Comment (godexsoft) [inline src/rpc/ngHandlers/LedgerData.cpp]: no need to try reuse, just create a new const here :) i understand this is old code but let's improve it

Comment (godexsoft) [reply src/rpc/ngHandlers/LedgerData.cpp]: you can even put it straight into the log line.. no real need for temporaries here

File: src/rpc/ngHandlers/LedgerData.cpp (hunk 6 above)
Comment (godexsoft) [reply src/rpc/ngHandlers/LedgerData.cpp]: or inline into log line even better

File: src/rpc/ngHandlers/LedgerData.h [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        return rpcSpec;
+    }
+
+    Result
+    process(Input input, Context const& ctx) const;
+
+private:
+    static uint32_t constexpr LIMITBINARY = 2048;
+    static uint32_t constexpr LIMITJSON = 256;
+    friend void
```
Comment (godexsoft) [inline src/rpc/ngHandlers/LedgerData.h:89]: needs newline before the function declaration

**Review by godexsoft — CHANGES_REQUESTED** [review]
Adding more small suggestions. Almost there :)

File: src/rpc/ngHandlers/LedgerData.cpp (hunk 4 above)
Comment (cindyyan317) [reply src/rpc/ngHandlers/LedgerData.cpp:92]: indeed . it should not be const in this case. will fix

File: src/rpc/ngHandlers/LedgerData.cpp [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+    auto const start = std::chrono::system_clock::now();
+    std::vector<Backend::LedgerObject> results;
+    if (input.diffMarker)
+    {
+        // keep the same logic as previous implementation
+        auto diff = sharedPtrBackend_->fetchLedgerDiff(*(input.diffMarker), ctx.yield);
+        std::vector<ripple::uint256> keys;
+        for (auto& [key, object] : diff)
+        {
+            if (!object.size())
```
Comment (cindyyan317) [inline src/rpc/ngHandlers/LedgerData.cpp:94]: Not really sure why the check is !object.size(), maybe just wanna the removed item? there is no document whatsoever regarding the "out_of_order". any idea?@godexsoft

Comment (godexsoft) [reply src/rpc/ngHandlers/LedgerData.cpp:94]: i suppose it is checking if the object is empty, whatever that means for this object type. i'd leave the logic as is if you copied it from the old impl 🤷‍♂️

**Review by godexsoft — APPROVED** [review]
LGTM 🚀

**Pull request merged**